packages/dropdown/src/dropdown.vue

```diff
@@ -0,0 +1,264 @@
+      type: Boolean,
+      default: true,
+    },
+    placement: {
+      type: String,
+      default: 'bottom',
+    },
+    showTimeout: {
+      type: Number,
+      default: 250,
+    },
+    hideTimeout: {
+      type: Number,
+      default: 150,
+    },
+    tabindex: {
+      type: Number,
+      default: 0,
+    },
+    effect: {
+      type: String,
+      default: 'light',
+    },
+  },
+  emits: ['visible-change', 'click', 'command'],
+  setup(props, { emit, slots }) {
+    const _instance = getCurrentInstance()
+    const { ELEMENT } = useDropdown()
+
+    const timeout = ref<Nullable<number>>(null)
+
+    const visible = ref(false)
+    watch(
+      () => visible.value,
+      val => {
+        if(val) triggerElmFocus()
+        if(!val) triggerElmBlur()
+        emit('visible-change', val)
+      },
+    )
+
+    const focusing = ref(false)
+    watch(
+      () => focusing.value,
+      val => {
+        const selfDefine = triggerElm.value
+        if (selfDefine) {
+          if (val) {
+            selfDefine.classList.add('focusing')
+          } else {
+            selfDefine.classList.remove('focusing')
+          }
+        }
+      },
+    )
+
+    const triggerVnode = ref<Nullable<VNode>>(null)
+    const triggerElm = computed<Nullable<HTMLButtonElement>>(() =>
+      !props.splitButton
+        ? triggerVnode.value?.el
+        : triggerVnode.value?.el.querySelector('.el-dropdown__caret-button'),
+    )
+
+    function handleClick() {
+      if (triggerElm.value?.disabled) return
+      if (visible.value) {
+        hide()
+      } else {
+        show()
+      }
+    }
+
+    function show() {
+      if (triggerElm.value?.disabled) return
+      timeout.value && clearTimeout(timeout.value)
+      timeout.value = window.setTimeout(() => {
+        visible.value = true
+      }, props.trigger === 'click' ? 0 : props.showTimeout)
+    }
+
+    function hide() {
+      if (triggerElm.value?.disabled) return
+      removeTabindex()
+      if (props.tabindex >=0) {
+        resetTabindex(triggerElm.value)
+      }
+      clearTimeout(timeout.value)
+      timeout.value = window.setTimeout(() => {
+        visible.value = false
+      }, props.trigger === 'click' ? 0 : props.hideTimeout)
+    }
+
+    function removeTabindex() {
+      triggerElm.value?.setAttribute('tabindex', '-1')
+    }
+
+    function resetTabindex(ele) {
+      removeTabindex()
+      ele?.setAttribute('tabindex', '0')
+    }
+
+    function triggerElmFocus() {
+      triggerElm.value?.focus?.()
+    }
+    function triggerElmBlur() {
+      triggerElm.value?.blur?.()
+    }
+
+    // for dom
+    Object.assign(_instance, {
+      handleClick,
+      hide,
+      resetTabindex,
+    })
+
+    const dropdownSize = computed(() => props.size || (ELEMENT || {}).size)
+    function commandHandler (...args) {
+      emit('command', ...args)
+    }
+    function initDom(instance) {
+      useDropdownDomEvent(instance, triggerElm.value, _instance)
```

Comment:
initDom is a callback. I don't think using hook in a non-setup function is good...

Reply:
OK, I will move it to  `dropdown-menu.vue`